internal/generator/generator_util.go

```diff
@@ -0,0 +1,238 @@
+	strContent = multiLineCommentExtracter.ReplaceAllString(strContent, "")
+
+	strContent = strings.ReplaceAll(strContent, "\r\n", "")
+	strContent = strings.ReplaceAll(strContent, "\n", "")
+
+	// Extract directives definition code blocks, each code block contains a list of directives definition
+	directiveDefBlocks := directivesDefBlockExtracter.FindAllStringSubmatch(strContent, -1)
+	// Iterate through every code block
+	for _, block := range directiveDefBlocks {
+		// The name of the directives block in source code, it may be used as the context
+		blockName := block[1]
+		// Extract directives and their attributes in the code block, the first dimension of attributesList
+		// is index of directive, the second dimension is index of attributes
+		attributesList := singleDirectiveExtracter.FindAllStringSubmatch(block[2], -1)
+		// Iterate through every directive
+		for _, attributes := range attributesList {
+			// Extract attributes from the directive
+			directiveName := strings.TrimSpace(attributes[1])
+			diretiveBitmasks := strings.Split(attributes[2], "|")
+			haveContext := false
+
+			for idx, bitmaskName := range diretiveBitmasks {
+				bitmaskGoName, found := ngxBitmaskNameToGo[strings.TrimSpace(bitmaskName)]
+				if !found {
+					return nil, fmt.Errorf("parsing directive %s, bitmask %s in source code not found in crossplane", directiveName, bitmaskName)
+				}
+				diretiveBitmasks[idx] = bitmaskGoName
+				if _, found := allDirectiveContexts[bitmaskGoName]; found {
+					haveContext = true
+				}
+			}
+
+			// If the directive doesn't have context in source code, maybe we still have a human-defined context for it.
+			// An example is directives in mgmt module, which was included in N+ R31
+			if !haveContext {
+				context, found := directiveBlock2Context[blockName]
+				if found {
+					diretiveBitmasks = append(diretiveBitmasks, context)
+				}
+			}
+
+			if bitmaskDefList, exist := directive2Defs[directiveName]; exist {
+				bitmaskDefList = append(bitmaskDefList, diretiveBitmasks)
+				directive2Defs[directiveName] = bitmaskDefList
+			} else {
+				directive2Defs[directiveName] = []bitDef{diretiveBitmasks}
+			}
+		}
+	}
+	return directive2Defs, nil
+}
+
+// Extract directives definitions from source code folder through regex.
+func getDirectiveDefFromFolder(path string) (map[string][]bitDef, error) {
+	directive2Defs := make(map[string][]bitDef, 0)
+
+	err := filepath.WalkDir(path, func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		// Check if the entry is a C/C++ file
+		// Some dynamic modules are written in C++, like otel
+		if !d.IsDir() && (strings.HasSuffix(path, ".c") || strings.HasSuffix(path, ".cpp")) {
+			dir2defsInFile, err := getDirectiveDefFromFile(path)
+			if err != nil {
+				return err
+			}
+			for directive, defsInFile := range dir2defsInFile {
+				if _, found := directive2Defs[directive]; !found {
+					directive2Defs[directive] = []bitDef{}
+				}
```

Comment:
Not necessary. You can `append()` a nil slice.

Reply:
Amazing... applied